### PR TITLE
Rendering Shapes without explicit bounds.

### LINKED ISSUE #4044: [Windows] Not rendering Line without setting Height and Width
### Description

Not rendering Line without setting Height and Width on Windows.
image

### Steps to Reproduce

1. Download [ShapesDemos.zip](https://github.com/dotnet/maui/files/7845444/ShapesDemos.zip)
2. Launch the App on Windows and navigate to the Line sample.
 

### Version with bug

Preview 11

### Last version that worked well

Preview 11

### Affected platforms

Windows

### Affected platform versions

Windows 10 1909

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #4051: [Android] Wrong size and layout issues using Line, Polygon, Polyline and Path
### Description

Wrong size and layout issues using Line, Polygon, Polyline and Path

image
image

### Steps to Reproduce

1. Download the sample [ShapesDemos.zip](https://github.com/dotnet/maui/files/7845660/ShapesDemos.zip)
2. Navigate to the Line, Polyline, etc samples. 

### Version with bug

Preview 11

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 11

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #6204: Line (horizontal or vertical) does not show
### Description

When trying to create a horizontal or vertical `Line`, it will not show. A workaround is f.ex. when creating a horizontal line is to set Y2="1" so that it gets some elevation and therefore is seemingly rendered.

This will not show:
```xaml
 <Line
    Stroke="Red"
    StrokeThickness="3"
    X2="350" />
```

But this will show:
```xaml
<Line
    Stroke="Red"
    StrokeThickness="3"
    X2="350"
    Y2="1" />
```

### Steps to Reproduce

1. Create a File -> New MAUI app
2. Paste in the first XAML snippet provided above.

### Version with bug

Release Candidate 1 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, I was *not* able test on other platforms

### Affected platform versions

Android 11 and up

### Did you find any workaround?

The second snippet was a workaround for Android.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by VSadov**
### Description of Change

Bounds of a Shape can be inferred from its shape and specifying that should not be required.
Currently not specifying Height and Width makes lines and other shapes not render.

### Issues Fixed

Fixes #4044
Fixes #6204
Fixes #4051

### CODE REVIEW

**Comment by VSadov** [comment]
== Before the change:
Before

**Comment by VSadov** [comment]
=== After the change:
After

**Comment by VSadov** [comment]
The fix also fixes other demo scenarios like  Polygon, Polyline and Path.

File: src/Controls/src/Core/Shapes/Shape.cs [hunk 1]
```diff
@@ -261,5 +262,24 @@ PathF IShape.PathForBounds(Graphics.Rect viewBounds)
 
 			return path;
 		}
+
+		protected override Size MeasureOverride(double widthConstraint, double heightConstraint)
+		{
+			var result = base.MeasureOverride(widthConstraint, heightConstraint);
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/Shapes/Shape.cs:268]: Nice. This is the important part that was missing.

File: src/Controls/src/Core/Shapes/Shape.cs [hunk 2]
```diff
@@ -261,5 +262,24 @@ PathF IShape.PathForBounds(Graphics.Rect viewBounds)
...
+			if (result.Width != 0 && result.Height != 0)
+			{
+				return result;
+			}
+
+			var boundsByFlattening = this.GetPath().Bounds.Size;
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/Shapes/Shape.cs:274]: In general everything looks really nice but there are a couple of wrong cases:

image
This ellipse should fill all the yellow space.

image
This should render a spiral path in the center.

Sometimes the bounds by flattening are wrong. For example, in the second case, Width and Height is zero.

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]

File: src/Controls/src/Core/Shapes/Shape.cs (hunk 2 above)
Comment (VSadov) [reply src/Controls/src/Core/Shapes/Shape.cs:274]: if the spiral path scenario is similar to one in repro project for #4044, then I believe the reason for the 0x0 size is that we do not invalidate the Height/Width request for the PolyLine when after replacing its Points. I think we should. 
If I do invalidation, then the spiral path renders correctly for me (at least in the repro project):

image

Comment (VSadov) [reply src/Controls/src/Core/Shapes/Shape.cs:274]: As for the ellipse scenario, I think my repro does not have an equivalent. There is a scenario with ellipses, but it is just a few simple fixed sized ellipse shapes and they show just fine. 

How can I get the AutoSize Shapes Gallery ?

Comment (jsuarezruiz) [reply src/Controls/src/Core/Shapes/Shape.cs:274]: Here you can find the AutoSize sample https://github.com/dotnet/maui/blob/034167fdbf18e87e366f75e78e4bef076c488b16/src/Controls/samples/Controls.Sample/Pages/Controls/ShapesGalleries/AutoSizeShapesGallery.xaml

Comment (jsuarezruiz) [reply src/Controls/src/Core/Shapes/Shape.cs:274]: Open the .NET MAUI Gallery and navigate to Controls -> Shapes. Then tap the "More samples" Button and select the "AutoSize Shapes Gallery".

Comment (VSadov) [reply src/Controls/src/Core/Shapes/Shape.cs]: There were two issues:
- `GetBoundsByFlattening` with default `0.01` flatness produces clearly incorrect results for ellipse curves. Using `flatness == 1` temporarily is a workaround, but there is likely a bug in `GetBoundsByFlattening`
- `MeasureOverride` on Shape should consider Aspect. That was missing

Comment (VSadov) [reply src/Controls/src/Core/Shapes/Shape.cs]: After the last changes:

image

image

**Review by jsuarezruiz — APPROVED** [review]

**Comment by VSadov** [comment]
Is this ok to merge? I can't tell if failures on `Android Device Tests Core (API 30)` are infrastructure issues or related to the change.

**Comment by Redth** [comment]
Running the one another time, but I suspect the last failure is unrelated.

**Comment by myroot** [comment]
No way to call measure function on handler side. So can't get Shapes size, Anyone tell me how to get Shape size on Handler

**Pull request merged**